relay-general/src/protocol/measurements.rs

```diff
@@ -0,0 +1,171 @@
+use regex::Regex;
+
+use crate::types::{Annotated, Error, FromValue, Object, Value};
+
+lazy_static::lazy_static! {
+    static ref MEASUREMENT_NAME: Regex = Regex::new("^[a-z0-9-._]+$").unwrap();
+}
+
+#[derive(Clone, Debug, Default, PartialEq, Empty, FromValue, ToValue, ProcessValue)]
+#[cfg_attr(feature = "jsonschema", derive(JsonSchema))]
+pub struct Measurement {
+    /// Value of observed measurement value
+    #[metastructure(required = "true", skip_serialization = "empty")]
+    pub value: Annotated<f64>,
+}
+
+#[derive(Clone, Debug, Default, PartialEq, Empty, ToValue, ProcessValue)]
+#[cfg_attr(feature = "jsonschema", derive(JsonSchema))]
+pub struct Measurements(#[metastructure(skip_serialization = "empty")] pub Object<Measurement>);
+
+impl FromValue for Measurements {
+    fn from_value(value: Annotated<Value>) -> Annotated<Self> {
+        let mut processing_errors = vec![];
+
+        let mut measurements = Object::<Measurement>::from_value(value).map_value(|measurements| {
+            Self(measurements.into_iter().fold(
```

Comment:
Isn't it possible to use `retain` here? This essentially destructs and then constructs a new map in a very indirect way (for instance, using `fold` rather than `filter_map`).

Retain, on the other hand, drops values in place, which is considerably faster.

Reply:
Update, and thanks for pointing it to me @dashed: There is no `retain` in `BTreeMap`. 

The idiomatic way to do this would be via `filter_map`, then: Return `Some((key, value))` if you want to keep the entry, or `None` if you want to drop it. The benefit of this is that the compiler can infer all types of your parameters so you do not need to annotate them. 

Your solution for collecting processing errors looks good.